File: 2024/day2/cpp/level1.cpp
#include <iostream>
#include <string>
#include <vector>
#include <fstream>

using namespace std;

vector<int> get_steps(vector<int> levels) {
    vector<int> steps;

    for (int i = 0 ; i < levels.size()-1 ; i++) {
        steps.push_back(levels[i+1]-levels[i]);
    }
    return steps;
}

bool is_safe(vector<int> steps) {
    int sign = 0;
    
    for (int step : steps) {
        if (abs(step)>3 || abs(step) < 1) {
            return false;
        }
        if (sign == 0) {
            sign = abs(step)/step;
        } else {
            if (abs(step)/step != sign) {
                return false;
            }
        }
    }
    return true;
}

int charInt(char c) {
    return c - '0';
}

vector<int> get_levels(string report) {
    vector<int> levels;

    for (char c : report) {
        if (c != ' ') levels.push_back(charInt(c));
    }
    return levels;
}

int main() {
    ifstream file("input.txt");
    string buffer;
    int n_safe = 0;
    if (!file.is_open()) {
        cerr << "Error while opening the file" << endl;
        return 1;
    }

    while (getline(file,buffer)) {
        vector<int> levels {get_levels(buffer)};
        vector<int> steps {get_steps(levels)};

        if (is_safe(steps)) n_safe ++;
    }
    cout << n_safe << endl;
    return 0;
}

File: 2024/day2/python/level1.py
import time

start = time.time()

with open("input.txt") as file:
    n_safe = 0

    for report in file.readlines():
        levels = list(map(int,report.split(" ")))
        sign = 0

        for i in range(len(levels)-1):
            step = levels[i+1]-levels[i]
            if not abs(step) in (1,2,3):
                break
            current_sign = step/abs(step)
            if sign == 0:
                sign = current_sign
            elif sign != current_sign:
                break


        else:
            n_safe += 1
print(n_safe)

print(time.time()-start,"s")